whatrecord/makefile.py

```diff
@@ -0,0 +1,611 @@
+        """
+        Load a Makefile from a filename.
+
+        Parameters
+        ----------
+        filename : pathlib.Path or str
+            The filename.
+
+        working_directory : pathlib.Path or str, optional
+            The working directory to use when evaluating the Makefile contents.
+            Assumed to be the directory in which the makefile is contained, but
+            this may be overridden.
+
+        keep_os_env : bool, optional
+            Keep environment variables in ``.env`` from outside of ``make``,
+            as in those present in ``os.environ`` when executing ``make``.
+
+        encoding : str, optional
+            String encoding to use.
+
+        Raises
+        ------
+        RuntimeError
+            If unable to run ``make`` and get information.
+
+        Returns
+        -------
+        makefile : Makefile
+            The makefile information.
+        """
+        with open(filename, "rt") as fp:
+            contents = fp.read()
+        return cls.from_string(
+            contents,
+            filename=filename,
+            working_directory=working_directory,
+            encoding=encoding,
+            keep_os_env=keep_os_env,
+        )
+
+    @staticmethod
+    def find_makefile(
+        file_or_directory: AnyPath,
+    ) -> pathlib.Path:
+        """
+        Find a Makefile given a path, if it exists.
+
+        Accepts either a direct path to a Makefile or a directory containing
+        a Makefile.
+
+        Parameters
+        ----------
+        file_or_directory : str or pathlib.Path
+            Makefile directory or direct path.
+
+        Raises
+        ------
+        FileNotFoundError
+            If the file or directory does not contain a ``Makefile``.
+        """
+        path = pathlib.Path(file_or_directory)
+
+        if path.is_dir():
+            path = path.resolve() / "Makefile"
+
+        if not path.exists():
+            raise FileNotFoundError(file_or_directory)
+
+        return path.resolve()
+
+
+@dataclass
+class Dependency:
+    """A single EPICS dependency (module) in the build system."""
+    #: The name of the dependency, as derived from the path name
+    name: str = ""
+    #: The name of the dependency, as derived from the build system variable
+    variable_name: Optional[str] = None
+    #: The absolute path to the dependency
+    path: pathlib.Path = field(default_factory=pathlib.Path)
+    #: The parsed Makefile information for this dependency
+    makefile: Makefile = field(default_factory=Makefile)
+    #: Modules (etc.) which depend on this instance
+    dependents: List[pathlib.Path] = field(default_factory=list)
+    #: Modules (etc.) are required for this instance
+    dependencies: List[pathlib.Path] = field(default_factory=list)
+    #: The "root" node.  ``None`` to refer to itself.
+    root: Optional[DependencyGroup] = field(default=None, metadata=apischema.metadata.skip)
+
+    @classmethod
+    def from_makefile(
+        cls,
```

Comment:
nitpick: this is a `@classmethod` that doesn't use the `cls` variable

Reply:
Not a nitpick - a bug.